File: Level 2/[3차] 압축/solution.js
function solution(msg) {
  const alphabet = [..."ABCDEFGHIJKLMNOPQRSTUVWXYZ"];
  const dict = {};

  alphabet.forEach((char, index) => (dict[char] = index + 1));

  const answer = [];

  while (true) {
    const keys = Object.keys(dict);

    if (keys.includes(msg)) {
      answer.push(dict[msg]);
      break;
    }

    for (let i = 1; i <= keys.length + 1; i++) {
      if (!keys.includes(msg.slice(0, i))) {
        answer.push(dict[msg.slice(0, i - 1)]);
        dict[msg.slice(0, i)] = keys.length + 1;
        msg = msg.slice(i - 1);
        break;
      }
    }
  }

  return answer;
}

File: Level 2/[3차] 압축/solution.py
def solution(msg):
    alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    dict = {key: value for key, value in zip(alphabet, list(range(1, 27)))}

    answer = []

    while True:
        if msg in dict:
            answer.append(dict[msg])
            break

        for i in range(1, len(msg) + 1):
            if msg[0:i] not in dict:
                answer.append(dict[msg[0 : i - 1]])
                dict[msg[0:i]] = len(dict) + 1
                msg = msg[i - 1 :]
                break

    return answer
